File: 11-container-with-most-water/container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        def amount(lo, hi) -> int:
            return min(height[lo],height[hi]) * (hi-lo)


        lo , hi = 0, len(height)-1
        maxSoFar = 0



        while lo < hi:
            # calculate water here
            maxSoFar = max(maxSoFar, amount(lo,hi))

            # Shift shorter wall
            if height[lo] <= height[hi]:
                lo += 1
            else:
                hi -= 1
        return maxSoFar

        

File: 11-container-with-most-water/container-with-most-water.dart
import 'dart:math';

class Solution {
  int maxArea(List<int> height) {

    int l = 0;
    int r = height.length-1;

    int water =0;
    int curWater = 0;

    while(l < r){
        curWater = (r-l) * min(height[l], height[r]);
        water = max(water, curWater);

        if (height[l] < height[r]){
            l++;
        }else{
            r--;
        }
    }
    return water;
  }
}